kernel/libs/jhash/src/lib.rs

```diff
@@ -0,0 +1,331 @@
+// SPDX-License-Identifier: MPL-2.0
+
+//! This module implements hash algorithms developed by Bob Jenkins.
+//!
+//! For further information, visit: www.burtleburtle.net/bob/hash/doobs.html
+//!
+//! The hash functions in this module facilitate the production of 32-bit values for hash table lookups.
+//! Although the [Linux kernel's jhash](https://github.com/torvalds/linux/blob/master/include/linux/jhash.h)
+//! slightly differs from the original version, this module reimplements the Linux kernel's version.
```

Comment:
I'm not familiar with the hash algorithms. The current implementation is good enough, simple and promising.

However, it's still not a complete hash map implementation, for example, we don't adjust the bucket size at runtime. As an alternative, I'm wondering if it's okay to just use the [`hashbrown`](https://github.com/rust-lang/hashbrown) crate, since it's under the `rust-lang` organization and is already used by the Rust standard library to provide `HashMap`/`HashSet` implementations:
> Since Rust 1.36, this is now the `HashMap` implementation for the Rust standard library. However you may still want to use this crate instead since it works in environments without `std`, such as embedded systems and kernels.


The downside might be that [it uses unsafe code](https://github.com/search?q=repo%3Arust-lang%2Fhashbrown+unsafe&type=code), so if we use it, we have to include it in our TCB. However, it is already a `std` dependency, and I believe [it will become part of `liballoc` in the future](https://github.com/rust-lang/rust/issues/27242) (we have to be patient), so I think it's still worth considering.

What's your opinion?

Reply:
> The Linux inet hashtable employs fine-grained, per-bucket locking

Actually, when I reread the PR description, I think there is a strong reason for us to reimplement the hash map. The locking problem may not be solvable if we simply use `HashMap`-like APIs.

Reply:
Actually, when I reread the PR description, I think there is a strong reason for us to reimplement the hash map. The locking problem may not be solvable if we simply use HashMap-like APIs.

Yes, I think that's the reason we need to reimplement hashmap by ourselves. The current HashMap APIs cannot use fine-grained lock.

> However, it's still not a complete hash map implementation, for example, we don't adjust the bucket size at runtime. 

Yes, I agree. An complete implementation still requires much more efforts. I just using the most naive implementation now. We can implement a more efficient hashmap in the future.

Reply:
Makes sense, thanks for the explanation!